cpp/src/arrow/compute/exec/aggregate_node.cc

```diff
@@ -283,28 +386,36 @@ class ScalarAggregateNode : public ExecNode, public TracedNode {
     DCHECK_EQ(input, inputs_[0]);
 
     auto thread_index = plan_->query_context()->GetThreadIndex();
-
-    ARROW_RETURN_NOT_OK(DoConsume(ExecSpan(batch), thread_index));
+    auto handler = [this, thread_index](const ExecBatch& full_batch,
```

Comment:
Why do we pass the handler here instead of just implementing the logic inside `HandleSegments`. I found the indirection here a bit confusing/unnecessary. Is that a specific reason to do it this way?

Reply:
This factoring of the code separates the handling of a single segment from the iteration over segments. I find it clear but don't mind changing if it would help. However, if the code is to be combined, I wouldn't do it into `HandleSegments` but just inline everything here.